lib/Glimmer.pm

```diff
@@ -0,0 +1,494 @@
+    run($verbose, "glimmer3 $glimmeropts $genetic_code_switch $contigs_file $tmp_model $tmp_prefix.pass_1");
+    
+    
+    
+    #+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+    #... Extract upstream regions and START codon counts
+    #-----------------------------------------------------------------------
+    my $tmp_predictions_pass_1 = "$tmp_prefix.pass_1.predict";
+    print STDERR "\nExtracting upstream regions and START counts from $tmp_predictions_pass_1\n"
+	if $verbose;
+    
+    open(PREDICT,  "<$tmp_predictions_pass_1")
+	|| die "Could not read-open $tmp_predictions_pass_1";
+    
+    my $tmp_upstream = "$tmp_prefix.upstream";
+    open(UPSTREAM, ">$tmp_upstream")
+	|| die "Could not write-open $tmp_upstream";
+    
+    my %start_counts;		#...Hash to hold START codon counts
+    $start_counts{0}   = 0;
+    $start_counts{atg} = 0;
+    $start_counts{gtg} = 0;
+    $start_counts{ttg} = 0;
+    
+    $contig_id = qq();
+    while(defined($entry = <PREDICT>)) {
+	chomp $entry;
+	
+	if ($entry =~ m/^>(\S+)/) {
+	    $contig_id = $1;
+	    next;
+	}
+	elsif ($entry =~ m/^(\S+)\s+(\d+)\s+(\d+)/) {
+	    my ($id, $beg, $end) = ($1, $2, $3);
+	    
+	    my $sign = ($beg < $end) ? +1 : -1;
+	    
+	    my $end_start   = $beg + $sign * 2;
+	    my $start_codon = &get_dna_seq($contig_id, $beg, $end_start, $len_of, $seq_of);
+	    
+	    ++$start_counts{0};
+	    ++$start_counts{$start_codon};
+	    
+	    my $up_id = $id;
+	    $up_id =~ s/^orf/ups/o;
+	    my $up_beg = $beg - $sign * 25;
+	    my $up_end = $beg - $sign;
+	    
+	    my $up_seq;
+	    if ($up_seq = &get_dna_seq($contig_id, $up_beg, $up_end, $len_of, $seq_of)) {
+		&display_id_and_seq($id, \$up_seq, \*UPSTREAM);
+	    }
+	}
+	else {
+	    die "Could not parse prediction $.: $entry";
+	}
+    }
+    
+    #+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+    #... Compute pseudocount-stabilized START codon frequencies
+    #-----------------------------------------------------------------------
+    my $atg_freq    = ($start_counts{atg} + 80) / ($start_counts{0} + 100);
+    my $gtg_freq    = ($start_counts{gtg} + 15) / ($start_counts{0} + 100);
+    my $ttg_freq    = ($start_counts{ttg} +  5) / ($start_counts{0} + 100);
+    my $start_freqs = "$atg_freq,$gtg_freq,$ttg_freq";
+    
+    print STDERR "start_freqs = $start_freqs\n\n" if $verbose;
+    
+    
+    #+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+    #... Build Position Weight Matrix for upstream regions using ELPH
+    #-----------------------------------------------------------------------
+    print STDERR "\nExtracting upstream motifs from $tmp_predictions_pass_1\n";
+    
+    my $motif_width = 6;
+    my $tmp_motifs = "$tmp_prefix.motifs";
+    system("elph $tmp_upstream LEN=$motif_width > $tmp_motifs")
+	&& ((-s $tmp_motifs)
+	    || die("Could not extract upstream motifs")
+	   );
```

Comment:
this logic is not very clear -- you would be better off just splitting this into two statements.
```
system(...);
die "Could not extract upstream motifs" unless -s $tmp_motifs;